apiserver/facades/agent/uniter/uniter.go

```diff
@@ -1837,37 +1847,57 @@ func (u *UniterAPI) NetworkInfo(args params.NetworkInfoParams) (params.NetworkIn
 		if err != nil {
 			return params.NetworkInfoResults{}, err
 		}
+
+		// Look for any relation specific egress subnets.
+		relEgress := state.NewRelationEgressNetworks(u.st)
+		egressSubnets, err := relEgress.Networks(rel.Tag().Id())
+		if err != nil && !errors.IsNotFound(err) {
+			return params.NetworkInfoResults{}, err
+		} else if err == nil {
+			cidrs := egressSubnets.CIDRS()
+			if len(cidrs) > 0 {
+				bindingsToEgressSubnets[endpoint.Name] = cidrs
+			}
+		}
+
 		boundSpace, err := unit.GetSpaceForBinding(endpoint.Name)
 		if err != nil && !errors.IsNotValid(err) {
 			return params.NetworkInfoResults{}, err
 		}
 		// If the endpoint for this relation is not bound to a space, or
-		// is bound to the default space, we need to set the address info
-		// for the default space to the correct cross model ingress address.
+		// is bound to the default space, we need to look up the ingress
+		// address info which is aware of cross model relations.
 		if boundSpace == environs.DefaultSpaceName || err != nil {
 			ru, err := u.getRelationUnit(canAccess, rel.Tag().String(), unitTag)
 			if err != nil {
 				return params.NetworkInfoResults{}, err
 			}
 			address, err := ru.IngressAddress()
-			logger.Debugf("address for relation %v is %v", rel.Tag().Id(), address)
-			spaces.Remove(environs.DefaultSpaceName)
-			defaultSpaceInfo = &network.NetworkInfo{
-				Addresses: []network.InterfaceAddress{{
-					Address: address.Value,
-				}},
+			if err != nil {
+				return params.NetworkInfoResults{}, err
 			}
+			bindingsToIngressAddress[endpoint.Name] = address.Value
 		}
 	}
 
 	networkInfos := machine.GetNetworkInfoForSpaces(spaces)
-	if defaultSpaceInfo != nil {
-		networkInfos[environs.DefaultSpaceName] = state.MachineNetworkInfoResult{
-			NetworkInfos: []network.NetworkInfo{*defaultSpaceInfo},
-		}
-	}
 	for binding, space := range bindingsToSpace {
-		result.Results[binding] = networkingcommon.MachineNetworkInfoResultToNetworkInfoResult(networkInfos[space])
+		// The binding address information based on link layer devices.
+		info := networkingcommon.MachineNetworkInfoResultToNetworkInfoResult(networkInfos[space])
+
+		// Set egress and ingress address information.
+		info.EgressSubnets = bindingsToEgressSubnets[binding]
+		// If there is no ingress address explicitly defined for a given binding,
```

Comment:
I don't get this. If we have room for multiple addresses, why include only one? Charms can just pick the first one if they want to.

Reply:
This was a carry over from older code. We now return all the addresses.

Reply:
It now adds all addresses into the ingress address slice